boutdata/collect.py

```diff
@@ -307,392 +239,823 @@ def getDataFile(i):
+
+    Parameters
+    ----------
+    i : int
+        Processor number being read from
+    datafile : DataFile
+        File to read from
+    varname : str
+        Name of variable to read
+    result : numpy.Array
+        Array in which to put the data
+    is_fieldperp : bool
+        Is this variable a FieldPerp?
+    grid_info : dict
+        dict of grid parameters
+    dimensions : tuple of str
+        Dimensions of the variable
+    tind : slice
+        Slice for t-dimension
+    xind : slice
+        Slice for x-dimension
+    yind : slice
+        Slice for y-dimension
+    zind : slice
+        Slice for z-dimension
+    xguards : bool
+        Include x-boundary cells at either side of the global grid?
+    yguards : bool
+        Include y-boundary cells at either end of the global grid?
+
+    Returns
+    -------
+    temp_yindex, var_attributes
+    """
+    ndims = len(dimensions)
+
+    # ndims is 0 for reals, and 1 for f.ex. t_array
+    if ndims == 0:
+        if i != 0:
+            # Only read scalars from file 0
+            return None, None
+
+        # Just read from file
+        result[...] = datafile.read(varname)
+        return None, None
+
+    if ndims > 4:
+        raise ValueError("ERROR: Too many dimensions")
+
+    if not any(dim in dimensions for dim in ("x", "y", "z")):
+        if i != 0:
+            return None, None
 
-    if not any(dim in dimensions for dim in ('x', 'y', 'z')):
         # Not a Field (i.e. no spatial dependence) so only read from the 0'th file
-        if 't' in dimensions:
-            if not dimensions[0] == 't':
+        if "t" in dimensions:
+            if not dimensions[0] == "t":
                 # 't' should be the first dimension in the list if present
                 raise ValueError(
-                    varname + " has a 't' dimension, but it is not the first dimension "
-                    "in dimensions=" + str(dimensions)
+                    "{} has a 't' dimension, but it is not the first dimension "
+                    "in dimensions={}".format(varname, dimensions)
                 )
-            data = f.read(varname, ranges = [tind] + (ndims - 1) * [None])
+            result[:] = datafile.read(varname, ranges=[tind] + (ndims - 1) * [None])
         else:
             # No time or space dimensions, so no slicing
-            data = f.read(varname)
-        if datafile_cache is None:
-            # close the DataFile if we are not keeping it in a cache
-            f.close()
-        return BoutArray(data, attributes=var_attributes)
+            result[...] = datafile.read(varname)
+        return None, None
+
+    nxpe = grid_info["nxpe"]
+    nype = grid_info["nype"]
+    mxsub = grid_info["mxsub"]
+    mysub = grid_info["mysub"]
+    mxg = grid_info["mxg"]
+    myg = grid_info["myg"]
+    yproc_upper_target = grid_info["yproc_upper_target"]
+
+    # Get X and Y processor indices
+    pe_yind = i // nxpe
+    pe_xind = i % nxpe
```

Comment:
Might be something we want to actually write from BOUT++?

Reply:
We actually do already, but only since July 2019.